### DISCUSSION TITLE
How to implement custom web page with  variable  text?

### DISCUSSION THREAD

**Discussion by Evgen2** [Hieromon/AutoConnect#382] [How can I?] [answered]
For example, I have some counter and want to show it on custom webpage, so that this page will show current counter value.

With usage of AutoConnectText I can use only static text and can't use pointer to some char buffer, that can be updated on webpage loading

**Comment by DanMan32** [comment]
I was successful by concatenating a quoted static string with a variable for the autoconnecftext.
The actual posting of the page is done in the "on" function related to the custom page.
I used the model used by HelloWorld which I believe is using the w/o JSON method.

The only problem I have is that in trying to use extra spaces in the static text, the extra spaces are trimmed to one space.

My static text is "Pool Temperature:      "
My variable is a float poolTemp.
The autoconnecttext string is "Pool Temperature: "+string(poolTemp)
Or something like that.  Can't recall at the moment what I used to convert float to string

**Reply by Evgen2** [reply]
"Or something like that" doesn't work :-(
There is "Simple" example with generating "by hands" html code with variable (current time), but it is incompatible with other pages mage by Autoconnect

**Comment by Hieromon** [comment, marked as answer]
@Evgen2 I'm not sure what the variable text you mentioned is expecting, but if it means that some string value (you try to keep it with the char pointer) will be reflected in your custom web page as is, AutoConnect does not support it. That is, you cannot share the value member of the AutoConnectText with a sketch variable (for example, it is globally declared). It must be assigned to the value member of AutoConnectText, and the procedure is implemented by the [**AutoConnect::on**](https://hieromon.github.io/AutoConnect/achandling.html#where-to-pick-up-the-values) handler as @DanMan32 mentioned.

If you want to reflect dynamic variables in your custom web page, its general processing structure looks like this:

```cpp
#include <WiFi.h>
#include <WebServer.h>
#include <AutoConnect.h>

const char VARTEXT[] PROGMEM = R"(
{
  "uri": "/vartext",
  "title": "Custom",
  "menu": true,
  "element": [
    {
      "name": "var",
      "type": "ACText",
      "format": "Counter value %s"
    }
  ]
}
)";

// A source of the variable text as like counter
int counter;

AutoConnect portal;

// Assign your intended sketch variable to the value member of AutoConnectText
String onPage(AutoConnectAux& varPage, PageArgument& args) {
  AutoConnectText&  var = varPage["var"].as<AutoConnectText>();
  var.value = String(++counter);  // The value member of AutoConnectText is always of type String.
  return String("");
}

void setup() {
  portal.load(VARTEXT);           // load the custom web page
  portal.on("/vartext", onPage);  // Register the custom web page handler
  counter = 0;
  portal.begin();
}

void loop() {
  portal.handleClient();
}
```

In the sketch above, the `int counter` increment is done in the `/vertex` page-handler, but of course, you can increment the counter in the `loop()`. Its variable has a global declaration.
